### PR TITLE
Change leads pipeline

### PR DESCRIPTION
**Pull request by SamardakOleh**

### CODE REVIEW

File: client/src/modules/lead/EditLead/EditLeadHeader/EditLeadPipelinePopover/EditLeadPipelinePopover.tsx [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+  private onFunnelSelect(funnel: Funnel) {
+    this.setState({
+      selectedFunnel: funnel,
+    }, () => {
+      this.props.loadPipelinePopoverStages(this.state.selectedFunnel._id);
+    });
+  }
+
+  private onStageSelect = (stageId: string) => {
```
Comment (valentyntu) [inline client/src/modules/lead/EditLead/EditLeadHeader/EditLeadPipelinePopover/EditLeadPipelinePopover.tsx]: maybe `handleStageSelect` will be a better name?

File: client/src/modules/lead/EditLead/EditLeadHeader/EditLeadPipelinePopover/EditLeadPipelinePopover.tsx [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+          <DropdownItem className={styles.option} onClick={this.onFunnelSelect.bind(this, funnel)} key={funnel._id}>
+            {funnel.name}
+          </DropdownItem>
+        );
+        })}
+      </DropdownMenu>
+    );
+  }
+
+  private onFunnelSelect(funnel: Funnel) {
```
Comment (valentyntu) [inline client/src/modules/lead/EditLead/EditLeadHeader/EditLeadPipelinePopover/EditLeadPipelinePopover.tsx]: maybe `handleFunnelSelect` will be a better name?

File: client/src/modules/lead/EditLead/EditLeadHeader/EditLeadPipelinePopover/EditLeadPipelinePopover.tsx [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+  private toggleDropdown = () => {
+    this.setState((prevState: State) => ({
+      isDropdownOpen: !prevState.isDropdownOpen,
+    }));
+  }
+
+  private togglePopover = () => {    
+    this.setState({  
+      isDropdownOpen: false,
```
Comment (valentyntu) [inline client/src/modules/lead/EditLead/EditLeadHeader/EditLeadPipelinePopover/EditLeadPipelinePopover.tsx]: these values are same as the initial state values, you should make `getInitialState()` and use it here and on state initilaization

File: client/src/modules/lead/EditLead/EditLeadHeader/EditLeadPipelinePopover/EditLeadPipelinePopover.tsx [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+  }
+}
+
+const mapStateToProps = (state: any) => ({
+  funnels: state.dashboard.funnels,
+  lead: state.dashboard.editLead.lead,
+  stages: state.dashboard.editLead.stagesForFunnel,
+});
+
+
```
Comment (valentyntu) [inline client/src/modules/lead/EditLead/EditLeadHeader/EditLeadPipelinePopover/EditLeadPipelinePopover.tsx]: redundant empty line

File: client/src/models/Lead.ts [hunk 5]
```diff
@@ -19,3 +19,18 @@ export default interface Lead {
...
+  owner: User;
+  visibility?: number;
+  name: string;
+  order: number;
+  contact?: Contact;
+  organization?: Organization;
+  custom: CustomField[];
+  timestamp: Date;
+  status: string;
+  notes: Note[];
```
Comment (alexanderko) [inline client/src/models/Lead.ts]: don't duplicate intercase definition

File: client/src/models/Stage.ts [hunk 6]
```diff
@@ -2,7 +2,15 @@ import Funnel from './Funnel';
 
 export default interface Stage {
    _id: string;
-  funnel: string | Funnel;
+  funnel: string;
+  name: string;
+  order: number;
+  timestamp: Date;
```
Comment (alexanderko) [inline client/src/models/Stage.ts]: don't duplicate intercase definition

File: client/src/modules/lead/EditLead/EditLeadHeader/EditLeadPipelinePopover/EditLeadPipelinePopover.tsx [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+  private isFunnelChangedAndStagesLoaded (props: Props): boolean {
+    return this.props.stages !== props.stages && this.props.stages.length !== 0;
+  }
+
+  private renderDropdownOptions() {
+    return (
+      <DropdownMenu>
+        {this.props.funnels.map((funnel: Funnel) => {
+        return (
```
Comment (alexanderko) [inline client/src/modules/lead/EditLead/EditLeadHeader/EditLeadPipelinePopover/EditLeadPipelinePopover.tsx]: remove `return`

File: client/src/modules/lead/EditLead/EditLeadHeader/EditLeadPipelinePopover/EditLeadPipelinePopover.tsx [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+          <DropdownItem className={styles.option} onClick={this.handleFunnelSelect.bind(this, funnel)} key={funnel._id}>
```
Comment (alexanderko) [inline client/src/modules/lead/EditLead/EditLeadHeader/EditLeadPipelinePopover/EditLeadPipelinePopover.tsx]: split in lines

File: client/src/modules/lead/EditLead/EditLeadHeader/EditLeadPipelinePopover/EditLeadPipelinePopover.tsx [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+  }
+
+  public componentWillReceiveProps(nextProps: Props) {
+    if (this.isFunnelChangedAndStagesLoaded(nextProps)) {
+      this.setState({
+        selectedStageId: nextProps.stages[0]._id,
+      });
+    } 
+
+    if (this.props.stages === nextProps.stages && this.state.selectedFunnel._id !== nextProps.lead.stage.funnel._id) {
```
Comment (alexanderko) [inline client/src/modules/lead/EditLead/EditLeadHeader/EditLeadPipelinePopover/EditLeadPipelinePopover.tsx]: extract method method

**Pull request merged**